revokefs/main.c

```diff
@@ -0,0 +1,576 @@
+  .setxattr = callback_setxattr,
+  .getxattr = callback_getxattr,
+  .listxattr = callback_listxattr,
+  .removexattr = callback_removexattr
+};
+
+enum {
+  KEY_HELP,
+};
+
+static void
+usage (const char *progname)
+{
+  fprintf (stdout,
+           "usage: %s basepath mountpoint [options]\n"
+           "\n"
+           "   Makes basepath visible at mountpoint such that files are read-only, directories are writable\n"
+           "\n"
+           "general options:\n"
+           "   -o opt,[opt...]     mount options\n"
+           "   -h  --help          print help\n"
+           "   --socket=fd         Pass in the socket fd\n"
+           "   --exit-with-fd=fd   Pass the read end of a pipe to monitor parent process's exit\n"
+           "   --backend           Run the backend instead of fuse\n"
+           "\n", progname);
+}
+
+static int
+revokefs_opt_proc (void *data,
+                   const char *arg,
+                   int key,
+                   struct fuse_args *outargs)
+{
+  (void) data;
+
+  switch (key)
+    {
+    case FUSE_OPT_KEY_NONOPT:
+      if (base_path == NULL)
+        {
+          base_path = g_strdup (arg);
+          return 0;
+        }
+      return 1;
+    case FUSE_OPT_KEY_OPT:
+      return 1;
+    case KEY_HELP:
+      usage (outargs->argv[0]);
+      exit (EXIT_SUCCESS);
+    default:
+      fprintf (stderr, "see `%s -h' for usage\n", outargs->argv[0]);
+      exit (EXIT_FAILURE);
+    }
+  return 1;
+}
+
+struct revokefs_config {
+  int socket_fd;
+  int exit_with_fd;
+  int backend;
+};
+
+#define REVOKEFS_OPT(t, p, v) { t, offsetof(struct revokefs_config, p), v }
+
+static struct fuse_opt revokefs_opts[] = {
+  REVOKEFS_OPT ("--socket=%i", socket_fd, -1),
+  REVOKEFS_OPT ("--exit-with-fd=%i", exit_with_fd, -1),
+  REVOKEFS_OPT ("--backend", backend, 1),
+
+  FUSE_OPT_KEY ("-h", KEY_HELP),
+  FUSE_OPT_KEY ("--help", KEY_HELP),
+  FUSE_OPT_END
+};
+
+int
+main (int argc, char *argv[])
+{
+  struct fuse_args args = FUSE_ARGS_INIT (argc, argv);
+  int res;
+  struct revokefs_config conf = { -1, -1 };
+
+  res = fuse_opt_parse (&args, &conf, revokefs_opts, revokefs_opt_proc);
+  if (res != 0)
+    {
+      fprintf (stderr, "Invalid arguments\n");
+      fprintf (stderr, "see `%s -h' for usage\n", argv[0]);
+      exit (EXIT_FAILURE);
+    }
+
+  if (base_path == NULL)
+    {
+      fprintf (stderr, "Missing basepath\n");
+      fprintf (stderr, "see `%s -h' for usage\n", argv[0]);
+      exit (EXIT_FAILURE);
+    }
+
+  basefd = openat (AT_FDCWD, base_path, O_RDONLY | O_NONBLOCK | O_DIRECTORY | O_CLOEXEC | O_NOCTTY);
+  if (basefd == -1)
+    {
+      perror ("opening basepath: ");
+      exit (EXIT_FAILURE);
+    }
+
+  if (conf.backend)
+    {
+      if (conf.socket_fd == -1)
+        {
+          fprintf (stderr, "No --socket passed, required for --backend\n");
+          exit (EXIT_FAILURE);
+        }
+
+      do_writer (basefd, conf.socket_fd, conf.exit_with_fd);
```

Comment:
Personally I think `--exit-with-fd` should be handled in this file, if you're doing it as a separate thread. It's unrelated to everything else `do_writer` does.